src/webviews/apps/shared/app.ts

```diff
@@ -0,0 +1,114 @@
+import { provide } from '@lit/context';
+import { html } from 'lit';
+import { property } from 'lit/decorators.js';
+import type { CustomEditorIds, WebviewIds, WebviewViewIds } from '../../../constants.views';
+import type { Deferrable } from '../../../system/function';
+import { debounce } from '../../../system/function';
+import type { WebviewFocusChangedParams } from '../../protocol';
+import { DidChangeWebviewFocusNotification, WebviewFocusChangedCommand, WebviewReadyCommand } from '../../protocol';
+import { GlElement } from './components/element';
+import { ipcContext, loggerContext, LoggerContext } from './context';
+import type { Disposable } from './events';
+import { HostIpc } from './ipc';
+
+export abstract class GlApp<
+	State extends { webviewId: CustomEditorIds | WebviewIds | WebviewViewIds; timestamp: number } = {
+		webviewId: CustomEditorIds | WebviewIds | WebviewViewIds;
+		timestamp: number;
+	},
+> extends GlElement {
+	@property({ type: String }) name!: string;
+	@property({ type: String }) placement: 'editor' | 'view' = 'editor';
+
+	@provide({ context: ipcContext })
+	protected _ipc!: HostIpc;
+
+	@provide({ context: loggerContext })
+	protected _logger!: LoggerContext;
+
+	@property({ type: Object })
+	state!: State;
+
+	private readonly disposables: Disposable[] = [];
+	private _focused?: boolean;
+	private _inputFocused?: boolean;
+	private _sendWebviewFocusChangedCommandDebounced!: Deferrable<(params: WebviewFocusChangedParams) => void>;
+
+	override connectedCallback() {
+		super.connectedCallback();
+
+		this._logger = new LoggerContext(this.name);
+		this._logger.log('connected');
+
+		//const themeEvent = computeThemeColors();
+		//if (this.onThemeUpdated != null) {
+		//this.onThemeUpdated(themeEvent);
+		//disposables.push(onDidChangeTheme(this.onThemeUpdated, this));
+		//}
+
+		this._ipc = new HostIpc(this.name);
+		this.disposables.push(
+			this._ipc.onReceiveMessage(msg => {
+				switch (true) {
+					case DidChangeWebviewFocusNotification.is(msg):
+						window.dispatchEvent(new CustomEvent(msg.params.focused ? 'webview-focus' : 'webview-blur'));
+						break;
+				}
+			}),
+			this._ipc,
+		);
+		this._ipc.sendCommand(WebviewReadyCommand, undefined);
+
+		this._sendWebviewFocusChangedCommandDebounced = debounce((params: WebviewFocusChangedParams) => {
+			this._ipc.sendCommand(WebviewFocusChangedCommand, params);
+		}, 150);
+
+		// Removes VS Code's default title attributes on <a> tags
+		document.querySelectorAll('a').forEach(a => {
+			if (a.href === a.title) {
+				a.removeAttribute('title');
+			}
+		});
```

Comment:
I wonder if we should just do this in `GlElement`? Though it could add up to a lot of DOM work.

Reply:
Ugh, I hate to need every component to do this, and vscode only does it once on load

Reply:
Wondering if we even need this in Shadow DOM